beamer/agent/chain.py

```diff
@@ -476,6 +489,117 @@ def maybe_invalidate(claim: Claim, context: Context) -> None:
+    l1_safety_factor = 1.25
+    limit = int(l1_gas_cost * l1_gas_price * l1_safety_factor)
+
+    # Agent is claimer
+    if claim.claimer == context.address:
+        reward = int(claim.latest_claim_made.challenger_stake_total)
+    else:
+        # Agent is challenger
+        reward = claim.get_challenger_stake(context.address)
+        last_challenger = claim.latest_claim_made.last_challenger
+        if last_challenger == context.address:
+            reward -= (
+                claim.latest_claim_made.challenger_stake_total
+                - claim.latest_claim_made.claimer_stake
+            )
+    return reward > limit
+
+
+def _timestamp_is_l1_finalized(
+    timestamp: Timestamp, context: Context, target_chain_id: ChainId
+) -> bool:
+    # The internal memory should not be none because at this point
+    # the ChainUpdated event must have arrived, otherwise there would be no request
+    target_chain_finality = context.finality_periods[target_chain_id]
+    return int(time.time()) > timestamp + target_chain_finality
+
+
+@dataclass
+class _ResolveInfo:
+    context_key: str
+    identifier: ClaimId | RequestId
+    proof_tx: HexBytes
+
+
+def maybe_resolve(claim: Claim, context: Context) -> bool:
+    request = context.requests.get(claim.request_id)
+
+    assert request is not None, "Active claim for non-existent request"
+
+    timestamp_finalized = False
+
+    if claim.invalidation_timestamp:
+        timestamp_finalized = _timestamp_is_l1_finalized(
+            claim.invalidation_timestamp, context, request.target_chain_id
+        )
+        assert claim.invalidation_tx is not None
+        info = _ResolveInfo("l1_invalidations", claim.id, claim.invalidation_tx)
+    elif request.fill_timestamp:
+        timestamp_finalized = _timestamp_is_l1_finalized(
+            request.fill_timestamp, context, request.target_chain_id
+        )
+        assert request.fill_tx is not None
+        info = _ResolveInfo("l1_resolutions", request.id, request.fill_tx)
+
+    if not timestamp_finalized:
+        return False
+
+    if not _l1_resolution_threshold_reached(claim, context):
+        return False
+
+    if info.identifier in getattr(context, info.context_key):
+        return False
+
+    future = context.task_pool.submit(
+        run_relayer_for_tx,
+        context.config.base_chain_rpc_url,
+        context.target_rpc_url,
+        context.source_rpc_url,
+        context.config.account.key,
+        info.proof_tx,
+    )
+
+    def on_future_done(f: Future) -> None:
+        try:
+            f.result()
+        except Exception as ex:
+            context.logger.error(L1_RESOLUTION_MESSAGES[info.context_key]["error"], ex=ex)
+        finally:
+            context_field = getattr(context, info.context_key)
+            del context_field[info.identifier]
+
+    future.add_done_callback(on_future_done)
+    context_field = getattr(context, info.context_key)
+    context_field[info.identifier] = future
```

Comment:
Suggestion:

We can unify `context.invalidations` and `context.l1_resolutions` into just `context.l1_resolutions` (which holds both request resolutions and fill invalidations) and just use the tx_hash as the key. This is unique anyway and every message can only be resolved once